modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/portlet/action/GetSocialTrafficSourcesMVCResourceCommand.java

```diff
@@ -122,21 +121,21 @@ private JSONArray _getReferringSocialMediaJSONArray(
 			return _jsonFactory.createJSONArray();
 		}
 
-		Stream<ReferringSocialMedia> stream = referringSocialMediaList.stream();
-
 		Comparator<ReferringSocialMedia> comparator = Comparator.comparingInt(
 			ReferringSocialMedia::getTrafficAmount);
 
-		return JSONUtil.putAll(
-			stream.filter(
-				referringSocialMedia ->
-					referringSocialMedia.getTrafficAmount() > 0
-			).sorted(
-				comparator.reversed()
-			).map(
-				referringSocialMedia -> referringSocialMedia.toJSONObject(
-					resourceBundle)
-			).toArray());
+		referringSocialMediaList = ListUtil.filter(
+			referringSocialMediaList,
+			referringSocialMedia ->
+				referringSocialMedia.getTrafficAmount() > 0);
+
+		referringSocialMediaList.sort(comparator.reversed());
```

Comment:
@slnn For use cases like "filter - sort - transform", please double check whether the order of these operations matters. For example, in this case, do we really need to filter first, and then sort, and then map? Is it possible to sort directly, and then filter when mapping? Also check similar cases in `GetDataMVCResourceCommand`.

Reply:
Hi @dantewang 

For the use case  "filter-sort-transform", I do not find out any other special reasons but one concern is that if we do filter first and then sort the list, we only need to sort the elements in the list which satisfy the condition of the filter, so if the filter action is faster than the sort with same amount elements then we will get better performance with keep the pattern, otherwise, we should choose sort -filter-tansform pattern.

By the way, keep the filter-sort-transform pattern, we can gain another benefit by using ListUtil.filter() that avoids creating a new List manually to contain the original list which may be an unmodifiable list that can not sort directly since ListUtil.filter() will handle it.

Lily

Reply:
@slnn OK then, yes, this is kind of hard to judge. Let's keep them for now, and optimize later if any performance issue is caught.